examples/larger-examples/datalog/subtyping.flix

```diff
@@ -0,0 +1,176 @@
+            Subtype(sub, super) :- Class(sub), Interface(super), Implementation(sub, super).
+            Subtype(sub, super) :- Class(sub), Class(super), ClassExtends(sub, super).
+            Subtype(sub, super) :- Interface(sub), Interface(super), InterfaceExtends(sub, super).
+            Subtype(sub, super) :- Subtype(sub, a), Subtype(a, super).
+
+            AvailableMethod(class, method) :- Method(t, method), Subtype(class, t).
+            NonSupertype(class)            :- Class(class), Class(otherClass), not Subtype(otherClass, class), if class != otherClass.
+            ExtendsFinalClass(sub, super)  :- FinalClass(super), Subtype(sub, super), if sub != super.
+        };
+        #(Subtype, AvailableMethod, NonSupertype, ExtendsFinalClass) -> rules <+> toPredicates(program)
+
+    def toPredicates(program: Program): #{ Class(String), FinalClass(String), Interface(String), ClassExtends(String, String),  InterfaceExtends(String, String), Implementation(String, String), Method(String, String) | r } =
+        let cs    = inject program.classes          into Class;
+        let fcs   = inject program.finalClasses     into FinalClass;
+        let is    = inject program.interfaces       into Interface;
+        let impls = inject program.implementations  into Implementation;
+        let ce    = inject program.classExtends     into ClassExtends;
+        let ies   = inject program.interfaceExtends into InterfaceExtends;
+        let ms    = inject program.methods          into Method;
+        cs <+> fcs <+> is <+> ce <+> impls <+> ies <+> ms
+
+    pub def subtypesDistinct(types: Program): Vector[(String, String)] =
+        let v = query subtypes(types) select (x, y) from Subtype(x, y);
+        Vector.filter(match (x, y) -> x != y, v)
+
+    pub def methodsOf(tpe: String, types: Program): Vector[String] =
+        query subtypes(types) select m from AvailableMethod(tpe, m)
+
+    pub def nonSupertypes(types: Program): Vector[String] =
+        query subtypes(types) select tpe from NonSupertype(tpe)
+
+    pub def finalClassExtensions(types: Program): Vector[(String, String)] =
+        query subtypes(types) select (sub, sup) from ExtendsFinalClass(sub, sup)
+
+    pub def isSubtype(subtype: String, possibleSupertype: String, types: Program): Bool =
+        let v = query subtypes(types) select (x, y) from Subtype(x, y);
+        not (Vector.filter(Eq.eq((subtype, possibleSupertype)), v) |> Vector.isEmpty)
+
+    pub def hasMethod(tpe: String, method: String, types: Program): Bool =
+        not (methodsOf(tpe, types) |> Vector.filter(Eq.eq(method)) |> Vector.isEmpty)
+
+    pub def isNonSupertype(tpe: String, types: Program): Bool =
+        not (nonSupertypes(types) |> Vector.filter(Eq.eq(tpe)) |> Vector.isEmpty)
+
+    pub def extendsFinalClass(tpe: String, types: Program): Bool =
+        not (finalClassExtensions(types) |> Vector.filter(match (sub, _) -> sub == tpe) |> Vector.isEmpty)
+
+
```

Comment:
Dropping all of the above queries, I think the following functions would be useful:

- `def hasField(p: Program, field: String): Bool`
- `def hasMethod(p: Program, method: String): Bool`
- `def illegalNonThrowable(p: Program, throws: Vector[...]): Vector[...]` // can only throw subtypes of Throwable.
- `def illegalExtendsFinalClass(p: Program, finalClasses: Vector[...]): Vector[...]` // cannot extend final classes.